File: PlusOne/PlusOne.cpp
#include <vector>
#include <iostream>
#include <assert.h>
#include <sstream>
#include <string>
#include <iterator>
#include <algorithm>

using namespace std;
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        
        std::vector<int> results;
        int carry = 1;
        int i = digits.size() - 1;

        // iterate over each digit, add carry at a time
        while (i >= 0)
        {
            auto result = digits[i] + carry;
            carry = result / 10;
            result %= 10;
            --i;
            results.push_back(result);
        }

        if (carry)
        {
            results.push_back(carry);
        }

        std::reverse(results.begin(), results.end());

        return results;
    }
};

bool equals(const vector<int> a, const vector<int> b)
{
    bool equal = true;
    
    if (a.size() != b.size())
    {
        return false;
    }

    for (size_t i = 0; i < a.size(); ++i)
    {
        if (a[i] != b[i])
            equal = false;
    }

    return equal;
}

int main()
{

    Solution sol;

    const vector<vector<int>> tests = {
        { 4, 3, 2, 1 },
        { 9, 9, 9 }
    };

    const vector<vector<int>> expected = {
        { 4, 3, 2, 2 },
        { 1, 0, 0, 0 }
    };

    int i = 0;
    for (auto test: tests)
    {
        auto soln = sol.plusOne(test);
        assert(equals(expected[i], soln));
        std::cout << "Test " << i << " passed.\n";
        i++;
    };

    return 0;
}

File: cvut/ch05/plus_one.py

def plus_one(n):

    n_result = []
    if len(n) == 0:
        return [1]

    carry = 1
    for i in range(len(n)):
        result = carry + n[len(n) - i - 1] % 10
        digit = result % 10
        carry = result // 10
        n_result.append(digit)
    if carry:
        n_result.append(carry)

    n_result.reverse()
    return n_result

def plus_one_2(n):

    if (len(n) == 0):
        return [1]
    
    n[len(n) - 1] += 1
    for i in reversed(range(len(n))):
        el = n[i]
        if el == 10 and i > 0:
            n[i] = 0
            n[i - 1] += 1
        if el == 10 and i == 0:
            n[i] = 1
            n.append(0)
    
    return n

print(plus_one([1,2,9]))
print(plus_one([9,9,9]))
print(plus_one([]))

inputs = [
    [1,2,9],
    [9,9,9],
    []
]

for i in inputs:
    print(plus_one_2(i))